apis/v1alpha2/gateway_types.go

```diff
@@ -0,0 +1,835 @@
+	// handshake for requests that match the hostname of the associated
+	// listener.
+	//
+	// References to a resource in different namespace are invalid UNLESS there
+	// is a ReferencePolicy in the target namespace that allows the certificate
+	// to be attached. If a ReferencePolicy does not allow this reference, the
+	// "ResolvedRefs" condition MUST be set to false for this listener with the
+	// "InvalidCertificateRef" reason.
+	//
+	// This field is required when mode is set to "Terminate" (default) and
+	// optional otherwise.
+	//
+	// CertificateRef can reference a standard Kubernetes resource, i.e. Secret,
+	// or an implementation-specific custom resource.
+	//
+	// Support: Core (Kubernetes Secrets)
+	//
+	// Support: Implementation-specific (Other resource types)
+	//
+	// +optional
+	CertificateRef *ObjectReference `json:"certificateRef,omitempty"`
+
+	// Options are a list of key/value pairs to give extended options
+	// to the provider.
+	//
+	// There variation among providers as to how ciphersuites are
+	// expressed. If there is a common subset for expressing ciphers
+	// then it will make sense to loft that as a core API
+	// construct.
+	//
+	// Support: Implementation-specific
+	//
+	// +optional
+	Options map[string]string `json:"options,omitempty"`
+}
+
+// TLSModeType type defines how a Gateway handles TLS sessions.
+//
+// +kubebuilder:validation:Enum=Terminate;Passthrough
+type TLSModeType string
+
+const (
+	// In this mode, TLS session between the downstream client
+	// and the Gateway is terminated at the Gateway.
+	TLSModeTerminate TLSModeType = "Terminate"
+	// In this mode, the TLS session is NOT terminated by the Gateway. This
+	// implies that the Gateway can't decipher the TLS stream except for
+	// the ClientHello message of the TLS protocol.
+	TLSModePassthrough TLSModeType = "Passthrough"
+)
+
+// ListenerRoutes defines which Routes may be attached to this Listener.
+type ListenerRoutes struct {
+	// Namespaces indicates which namespaces Routes may be attached to this
+	// Listener from. This is restricted to the namespace of this Gateway by
+	// default.
+	//
+	// Support: Core
+	//
+	// +optional
+	// +kubebuilder:default={from: Same}
+	Namespaces *RouteNamespaces `json:"namespaces,omitempty"`
+
+	// Kinds specifies the groups and kinds of Routes that are allowed to bind
+	// to this Gateway Listener. When unspecified or empty, the kinds of Routes
+	// selected are determined using the Listener protocol.
+	//
+	// A RouteGroupKind MUST correspond to kinds of Routes that are compatible
+	// with the application protocol specified in the Listener's Protocol field.
+	// If an implementation does not support or recognize this resource type, it
+	// MUST set the "ResolvedRefs" condition to false for this Listener with the
+	// "InvalidRoutesRef" reason.
+	//
+	// Support: Core
+	//
+	// +optional
+	// +kubebuilder:validation:MaxItems=8
+	Kinds []RouteGroupKind `json:"kinds,omitempty"`
```

Comment:
So all routes defined in the target namespace `granted that matches the namespace specification same/selector etc` will be automatically bound to this listener?

Is this how we expect users to compartmentalize routes.. by namespace?

Reply:
On the Gateway side we're just describing the kinds/namespaces of Routes of that may be attached to a Gateway. If nothing is specified, Routes in the same namespace with a matching protocol are trusted to attach to the Gateway. So this is all about defining where Routes _can_ attach from.

The primary attachment mechanism is from Routes - they directly reference the Gateways they want to attach to. A Route cannot be attached to a Gateway without directly referencing it.

Reply:
While this is functional. It is not what usually people expect. The current un-written-rules is using selectors (and netpol introduced namespace selectors). usually the parent selects children. Not the other way around. So this needs to be super crisp to users.

Reply:
Yep, agree with that. We did start with a selector from Gateways to Routes. This is a major change in v1alpha2. The fundamental issue with selectors is that they gave Gateway admins a false sense of control. Even when they specified a Route selector, it was still the Route owners that had control of Gateways they were attached to, simply by adjusting the labels on their Routes. The only true filtering control Gateway admins had is still in place namespaces, route kinds, and hostnames.

Reply:
There's more info/rationale in the GEP that proposed this change: https://gateway-api.sigs.k8s.io/geps/gep-724/.